src/game/AI/ScriptDevAI/scripts/northrend/naxxramas/instance_naxxramas.cpp

```diff
@@ -775,6 +864,150 @@ InstanceData* GetInstanceData_instance_naxxramas(Map* pMap)
     return new instance_naxxramas(pMap);
 }
 
+struct Location3DPoint
+{
+    float x, y, z;
+};
+
+static const Location3DPoint gargoyleResetCoords = {2963.f, -3476.f, 297.6f};
+
+enum
+{
+    SAY_GARGOYLE_NOISE      = -1533160, // %s emits a strange noise.
+
+    SPELL_STONEFORM         = 29154,
+    SPELL_STEALTH_DETECTION = 18950,
+    SPELL_STONESKIN         = 28995,
+    SPELL_ACID_VOLLEY       = 29325,
+    SPELL_ACID_VOLLEY_25    = 54714,
+};
+
+struct npc_stoneskin_gargoyleAI : public ScriptedAI
+{
+    npc_stoneskin_gargoyleAI(Creature* pCreature) : ScriptedAI(pCreature)
+    {
+        m_creature->GetCombatManager().SetLeashingCheck([&](Unit*, float x, float y, float z)
+        {
+            return x > gargoyleResetCoords.x && y > gargoyleResetCoords.y && z > gargoyleResetCoords.z;
+        });
+        Reset();
+    }
+
+    uint32 acidVolleyTimer;
+    bool canCastVolley;
+
+    void Reset() override
+    {
+        acidVolleyTimer = 4000;
+        canCastVolley = false;
+        TryStoneForm();
+
+        DoCastSpellIfCan(m_creature, SPELL_STEALTH_DETECTION, CAST_TRIGGERED | CAST_AURA_NOT_PRESENT);
+    }
+
+    void TryStoneForm()
+    {
+        if (m_creature->GetDefaultMovementType() == IDLE_MOTION_TYPE)
+        {
+            if (DoCastSpellIfCan(m_creature, SPELL_STONEFORM, CAST_TRIGGERED | CAST_AURA_NOT_PRESENT) == CAST_OK)
+            {
+                m_creature->SetFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_UNINTERACTIBLE);
+                m_creature->SetImmuneToPlayer(true);
+            }
+        }
+    }
+
+    void JustReachedHome() override
+    {
+        TryStoneForm();
+    }
+
+    void MoveInLineOfSight(Unit* pWho) override
+    {
+        if (m_creature->HasFlag(UNIT_FIELD_FLAGS, UNIT_FLAG_UNINTERACTIBLE))
```

Comment:
Why is this script needed?

Reply:
It wasn't implemented in EAI last I checked, but I'm not sure